Open A Connection

DeKaylaPhillips opened the pull request:
<!--
     Before submitting a Pull Request, please ensure you've done the following:
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.
-->
# Open A Connection

## Related Tickets & Documents

- https://trello.com/c/OoEBcy2I/38-open-a-connection

![image](https://github.com/DeKaylaPhillips/Echo-Server/assets/91583190/f8514642-2994-43bb-ae8b-0a7b33ecf999)

## Type of PR <!-- (check all applicable) -->

- [X] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This branch updates the source code used to create the server. The branch removes the previous source code using the external Express.js library, and replaces the previous implementation with a new implementation that utilizes the NET Node.js module objects to create a TCP server. The server socket listens for TCP connections on Port 3000 at the localhost IP address. 

<img width="486" alt="image" src="https://github.com/DeKaylaPhillips/Echo-Server/assets/91583190/556deb59-0160-41c3-a25e-ed71cf3f2300">

## Added/updated tests?

- [X] Yes
- [ ] No, and this is why: ___
- [ ] I need help with writing tests

Commit:
Refactor server module

Imports updated to utilize the NET Node.js module instead of Express.js module. The source code has been updated to implement the NET library, and was used to create and run the TCP server.

Naomi-Dennis commented on src/server.ts:
```diff
@@ -1,7 +1,10 @@
+const server = net.createServer((socket: net.Socket): void => {})
+
+server.listen(port, host, (): void => {
+  console.log(`TCP server has established a connection at http://${host}:${port}/.`)
```
Indeed, it is important not to test code we don't control. 

But what about this console log statement? 

I suggest writing a test for this!

DeKaylaPhillips replied:
The file this references was added to its own 'server' directory and can be found in `src/server/server.ts`, and the mock test for the console statement has been created and can be found in `tests/console/console.ts`. Thank you for the help! :)

Commit:
Update package.json

Commit:
Refactor server module

Added directories to organize each module. Added run file to run the server, refactored the server to utilize a class-based structure, and added mock to test the behavior of the server and its console response.

Commit:
Update eslint config file

Added maximum length rule.

Commit:
Refactor and cleanup

Refactored and cleaned server source code and tests to abide by new lint rule. Shortened console statement and test description.

Naomi-Dennis commented on tests/console/console.spec.ts:
```diff
@@ -0,0 +1,45 @@
+  }
+})
+
+describe('Console', () => {
```
This should be changed to 'Server' since we're testing the Server behavior, not the console behavior

Naomi-Dennis replied:
Also, the filename should change. The console is something created by JavaScript. It's not code we control. It outputting to the user as a consequence of the connection opening is what's being tested; _not_ the console itself.

I suggest putting this in a `server` folder instead of a `console` folder and renaming the file to `server.spec.ts`

DeKaylaPhillips replied:
The directory, file, and test name were renamed to 'Server'. :)

Naomi-Dennis commented on tests/console/console.spec.ts:
```diff
@@ -0,0 +1,45 @@
+    server = new Server(port, host)
+  })
+
+  test('logs a message to the terminal when a connection establishes', () => {
+    logMock = jest.spyOn(console, 'log').mockImplementation()
+    server.openTCPConnection()
+    expect(logMock).toHaveBeenCalledTimes(1)
+    expect(logMock).toHaveBeenCalledWith(
+      `TCP server at http://${host}:${port}/.`
+    )
+    logMock.mockRestore()
+  })
+})
```
🎉

Commit:
Clean lint errors and rename server dir and file

Naomi-Dennis commented on tests/controller/index.spec.ts:
```diff
@@ -16,7 +16,9 @@ describe('Index Controller', () => {
 describe('Message Controller', () => {
   test("responds with a user'/s message", () => {
-    const mockRequest: Partial<Request<any>> = { params: { message: 'I am a user!' } }
+    const mockRequest: Partial<Request<any>> = {
+      params: { message: 'I am a user!' }
+    }
```
This file and folder can be removed since the controller was removed from implementation!

DeKaylaPhillips replied:
Removing the files/folder! :)

Commit:
Remove controller files/directories
